proxy/envoy/config.go

```diff
@@ -124,83 +112,126 @@ func Generate(instances []*model.ServiceInstance, services []*model.Service, mes
-			LbType:           LbTypeRoundRobin,
-			Hosts:            []Host{{URL: fmt.Sprintf("tcp://%s:%d", "127.0.0.1", port)}},
+	for _, instance := range instances {
+		port := instance.Endpoint.Port
+		ensure(port)
+		ports[port.Port].instances[port.Protocol] = true
+	}
+
+	for _, svc := range services {
+		for _, port := range svc.Ports {
+			ensure(port)
+			ports[port.Port].services[port.Protocol] = append(ports[port.Port].services[port.Protocol], &model.Service{
+				Name:      svc.Name,
+				Namespace: svc.Namespace,
+				Ports:     []model.Port{port},
+			})
 		}
+	}
 
+	for port, lst := range ports {
 		listener := Listener{
 			Port:       port,
 			BindToPort: false,
 		}
 
+		ingressClusterName := fmt.Sprintf("%s%d", IngressClusterPrefix, port)
+		if len(lst.instances) > 0 {
+			clusters = append(clusters, Cluster{
+				Name:             ingressClusterName,
+				Type:             "static",
+				ConnectTimeoutMs: 1000,
+				LbType:           LbTypeRoundRobin,
+				Hosts:            []Host{{URL: fmt.Sprintf("tcp://%s:%d", "127.0.0.1", port)}},
+			})
+		}
+
 		// These two filters operate at different levels of the network stack.
 		// In practice, no port has two protocols used in the same set of instances, but we
 		// should be careful with not stepping on our feet.
-		if protocols[model.ProtocolTCP] {
+		// The order of the filter insertion is important.
+		if lst.instances[model.ProtocolTCP] {
 			listener.Filters = append(listener.Filters, NetworkFilter{
 				Type: "read",
 				Name: "tcp_proxy",
 				Config: NetworkFilterConfig{
-					Cluster:    cluster.Name,
+					Cluster:    ingressClusterName,
 					StatPrefix: "ingress_tcp",
 				},
 			})
 		}
 
-		if protocols[model.ProtocolGRPC] || protocols[model.ProtocolHTTP2] || protocols[model.ProtocolHTTP] {
+		// HTTP-level filter is shared between ingress and egress traffic.
+		// The routing decision is based on the virtual host.
+		hosts := make([]VirtualHost, 0)
+		for _, proto := range []model.Protocol{model.ProtocolHTTP, model.ProtocolHTTP2, model.ProtocolGRPC} {
+			for _, svc := range lst.services[proto] {
+				// TODO: support all variants for services: name.<my namespace>, name.namespace.svc.cluster.local
+				hosts = append(hosts, VirtualHost{
+					Name:    svc.String(),
+					Domains: []string{svc.Name, svc.Name + "." + svc.Namespace},
+					Routes:  []Route{{Cluster: EgressClusterPrefix + svc.String()}},
+				})
+			}
+		}
+		sort.Sort(HostsByName(hosts))
+
+		// The order of the hosts is important: host * is a fallback
+		if lst.instances[model.ProtocolHTTP] || lst.instances[model.ProtocolHTTP2] || lst.instances[model.ProtocolGRPC] {
+			hosts = append(hosts, VirtualHost{
+				Name:    "ingress_http",
+				Domains: []string{"*"},
+				Routes:  []Route{{Cluster: ingressClusterName}},
+			})
+		}
+
+		// TODO: HTTPS protocol ingress and egress configuration
+
+		if len(hosts) > 0 {
 			listener.Filters = append(listener.Filters, NetworkFilter{
 				Type: "read",
```

Comment:
Its fine if its not there in this iteration. But lets just mark it as a TODO, so that we don't miss it and that everyone knows about the issue.